src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs

```diff
@@ -0,0 +1,210 @@
+                /// <summary>
+                /// Every walker is walking a specific symbol, and can potentially walk each symbol multiple times
+                /// to get to a stable state. Each of these symbols gets a single global state slot, which this
+                /// dictionary keeps track of. These slots correspond to indexes into <see cref="_walkerStates"/>.
+                /// </summary>
+                private readonly PooledDictionary<Symbol, int> _symbolToSlot = PooledDictionary<Symbol, int>.GetInstance();
+                private int _currentWalkerSlot = -1;
+                private Symbol _initialSymbol = null;
+#if DEBUG
+                private Symbol _currentSymbol = null;
+#endif
+
+                internal SnapshotManager ToManagerAndFree()
+                {
+                    Debug.Assert(_currentWalkerSlot == -1, "Attempting to finalize snapshots before all walks completed");
+                    Debug.Assert(_symbolToSlot.Count == _walkerStates.Count);
+                    _symbolToSlot.Free();
+                    return new SnapshotManager(
+                        _walkerStates.ToImmutableAndFree(),
+                        _incrementalSnapshots.ToImmutable());
+                }
+
+                internal void EnterNewWalker(Symbol symbol)
+                {
+#if DEBUG
+                    Debug.Assert(symbol is object);
+                    Debug.Assert((_currentSymbol is null && _currentWalkerSlot == -1 && _initialSymbol is null) ||
+                                 symbol.ContainingSymbol.Equals(_currentSymbol));
+                    _currentSymbol = symbol;
+#endif
+                    _initialSymbol ??= symbol;
+                    var previousSlot = _currentWalkerSlot;
+
+                    // Because we potentially run multiple passes, we
+                    // need to make sure we use the same final shared
+                    // state for following passes.
+                    if (_symbolToSlot.TryGetValue(symbol, out var slot))
+                    {
+                        _currentWalkerSlot = slot;
+                    }
+                    else
+                    {
+                        _currentWalkerSlot = _symbolToSlot.Count;
+                        _symbolToSlot.Add(symbol, _currentWalkerSlot);
+                    }
+                }
+
+                internal void ExitWalker(SharedWalkerState stableState, Symbol symbol)
+                {
+#if DEBUG
+                    Debug.Assert(_currentSymbol.Equals(symbol));
+                    _currentSymbol = symbol.ContainingSymbol;
+#endif
+                    _walkerStates.SetItem(_currentWalkerSlot, stableState);
+                    if (_initialSymbol.Equals(symbol))
+                    {
+                        Debug.Assert(_currentWalkerSlot == 0);
+                        _currentWalkerSlot = -1;
```

Comment:
>_currentWalkerSlot = -1; [](start = 24, length = 24)

It feels like we would want to reset _currentSymbol to null as well.

Reply:
And _initialSymbol   too.

---
In reply to: [289559721](https://github.com/dotnet/roslyn/pull/35850#discussion_r289559721) [](ancestors = 289559721)